source/main/gfx/GfxActor.cpp

```diff
@@ -1202,3 +1333,992 @@ void RoR::GfxActor::CalculateDriverPos(Ogre::Vector3& out_pos, Ogre::Quaternion&
+        }
+        //normalize
+        vdir = vdir / vlen;
+        prop.beacon_flare_billboard_scene_node[0]->setPosition(prop.beacon_light[0]->getPosition() - vdir * 0.1);
+        bool visible = false;
+        if (prop.beacon_light_rotation_angle[0] > 1.0)
+        {
+            prop.beacon_light_rotation_angle[0] = 0.0;
+            visible = true;
+        }
+        visible = visible && enableAll;
+        prop.beacon_light[0]->setVisible(visible);
+        prop.beacon_flare_billboard_scene_node[0]->setVisible(visible);
+    }
+    if (prop.beacontype == 'R' || prop.beacontype == 'L')
+    {
+        Vector3 mposition = nodes[prop.noderef].AbsPosition + prop.offsetx * (nodes[prop.nodex].AbsPosition - nodes[prop.noderef].AbsPosition) + prop.offsety * (nodes[prop.nodey].AbsPosition - nodes[prop.noderef].AbsPosition);
+        //billboard
+        Vector3 vdir = mposition - gEnv->mainCamera->getPosition();
+        float vlen = vdir.length();
+        if (vlen > 100.0)
+        {
+            prop.beacon_flare_billboard_scene_node[0]->setVisible(false);
+            return;
+        }
+        //normalize
+        vdir = vdir / vlen;
+        prop.beacon_flare_billboard_scene_node[0]->setPosition(mposition - vdir * 0.1);
+    }
+    if (prop.beacontype == 'w')
+    {
+        Vector3 mposition = nodes[prop.noderef].AbsPosition + prop.offsetx * (nodes[prop.nodex].AbsPosition - nodes[prop.noderef].AbsPosition) + prop.offsety * (nodes[prop.nodey].AbsPosition - nodes[prop.noderef].AbsPosition);
+        prop.beacon_light[0]->setPosition(mposition);
+        prop.beacon_light_rotation_angle[0] += dt * prop.beacon_light_rotation_rate[0];//rotate baby!
+        //billboard
+        Vector3 vdir = mposition - gEnv->mainCamera->getPosition();
+        float vlen = vdir.length();
+        if (vlen > 100.0)
+        {
+            prop.beacon_flare_billboard_scene_node[0]->setVisible(false);
+            return;
+        }
+        //normalize
+        vdir = vdir / vlen;
+        prop.beacon_flare_billboard_scene_node[0]->setPosition(mposition - vdir * 0.1);
+        bool visible = false;
+        if (prop.beacon_light_rotation_angle[0] > 1.0)
+        {
+            prop.beacon_light_rotation_angle[0] = 0.0;
+            visible = true;
+        }
+        visible = visible && enableAll;
+        prop.beacon_light[0]->setVisible(visible);
+        prop.beacon_flare_billboard_scene_node[0]->setVisible(visible);
+    }
+}
+
+void RoR::GfxActor::UpdateProps(float dt, bool is_player_actor)
+{
+    using namespace Ogre;
+
+    if (m_simbuf.simbuf_beaconlight_active != m_beaconlight_active)
+    {
+        m_beaconlight_active = m_simbuf.simbuf_beaconlight_active;
+        this->SetBeaconsEnabled(m_beaconlight_active);   
+    }
+
+    NodeData* nodes = this->GetSimNodeBuffer();
+
+    for (prop_t& prop: m_props)
+    {
+        if (!prop.scene_node)
+            continue;
+
+        // -- quick ugly port from `Actor::NotifyCameraChanged()`~ 06/2018
```

Comment:
What if the prop is a `scene_node`?

Reply:
> What if the prop is a scene_node?

I don't understand the question.

The `if (prop.scene_node == nullptr)` comes from the original code, somebody needs to fix it - no scenenode means setup failed and the prop should never make it to `m_props`.

Reply:
I think the `// -- quick ugly port from `Actor::NotifyCameraChanged()`~ 06/2018` needs to always run.

Reply:
Eh, that's just one of many things which need to be researched and cleaned up. I work in small iterations, first I cut+paste code with all quirks, then test, then I revisit it later. And everything related to the `AsyncScene` rewrite is in the 'pasted with quirks' stage.

Reply:
Yeah, but those subtle code changes are gonna bite us in the ass later on. :smile:

Reply:
That's a topic for a long forum thread, bugs will always happen, but short commits help pinpointing them, even if it means messier code in the meantime (that's what comments are for). I wouldn't stress causing a few visual glitches, what troubles me in the long run are physics/collisions/scripting.